mailpoet/lib/AdminPages/Pages/AutomationEditor.php

```diff
@@ -0,0 +1,69 @@
+<?php declare(strict_types=1);
+
+namespace MailPoet\AdminPages\Pages;
+
+use DateTimeImmutable;
+use MailPoet\AdminPages\PageRenderer;
+use MailPoet\Automation\Engine\Storage\WorkflowStorage;
+use MailPoet\Automation\Engine\Workflows\Step;
+use MailPoet\Automation\Engine\Workflows\Workflow;
+use MailPoet\WP\Functions as WPFunctions;
+
+class AutomationEditor {
+  /** @var WorkflowStorage */
+  private $workflowStorage;
+
+  /** @var PageRenderer */
+  private $pageRenderer;
+
+  /** @var WPFunctions */
+  private $wp;
+
+  public function __construct(
+    WorkflowStorage $workflowStorage,
+    PageRenderer $pageRenderer,
+    WPFunctions $wp
+  ) {
+    $this->workflowStorage = $workflowStorage;
+    $this->pageRenderer = $pageRenderer;
+    $this->wp = $wp;
+  }
+
+  public function render() {
+    // Gutenberg styles
+    $this->wp->wpEnqueueStyle('wp-edit-post');
+    $this->wp->wpEnqueueStyle('wp-format-library');
+    $this->wp->wpEnqueueMedia();
+
+    $id = isset($_GET['id']) ? (int)$_GET['id'] : null;
+
+    $workflow = $id ? $this->workflowStorage->getWorkflow($id) : null;
+    $this->pageRenderer->displayPage('automation/editor.html', [
+      'workflow' => $workflow ? $this->buildWorkflow($workflow) : null,
+      'sub_menu' => 'mailpoet-automation',
+      'api' => [
+        'root' => rtrim($this->wp->escUrlRaw($this->wp->restUrl()), '/'),
+        'nonce' => $this->wp->wpCreateNonce('wp_rest'),
+      ],
+    ]);
+  }
+
+  private function buildWorkflow(Workflow $workflow): array {
+    return [
```

Comment:
`Workflow::toArray()` does not work here?

Reply:
Not really. It's because `Workflow::toArray()` is for internal data saving, while here we're serving it to the output. What we may do here is to reuse a `GET /workflows` endpoint implementation internally, rather than duplicating it. Ultimately, it's what that is — just an optimization to avoid one extra request.